src/main/java/application/MenuController.java

```diff
@@ -87,34 +120,42 @@ private SessionBean getNextMenu(MenuSession menuSession, boolean redrawing) thro
             return menuSession.startFormEntry(sessionRepo);
         }
         else{
-            MenuResponseBean menuResponseBean = new MenuResponseBean();
-            menuResponseBean.setSessionId(menuSession.getSessionId());
+            MenuSessionBean menuResponseBean = new MenuSessionBean();
+
             // We're looking at a module or form menu
             if(nextScreen instanceof MenuScreen){
-                MenuScreen menuScreen = (MenuScreen) nextScreen;
-                menuResponseBean.setMenuType(Constants.MENU_MODULE);
-                menuResponseBean.setOptions(getMenuRows(menuScreen));
+                menuResponseBean = generateMenuScreen((MenuScreen) nextScreen);
             }
             // We're looking at a case list or detail screen (probably)
             else if (nextScreen instanceof EntityScreen) {
-                EntityScreen entityScreen = (EntityScreen) nextScreen;
-                menuResponseBean.setMenuType(Constants.MENU_ENTITY);
-                menuResponseBean.setOptions(getMenuRows(entityScreen.getCurrentScreen()));
+                menuResponseBean = generateEntityScreen((EntityScreen) nextScreen);
+            } else{
+                throw new Exception("What screen are we on? " + nextScreen);
```

Comment:
nit: might be an easier error to read if it's `Unable to determine next screen: `